folly/SocketPortability.h

```diff
@@ -0,0 +1,148 @@
+#include <io.h>
+#undef __STDC__
+#else
+#include <io.h>
+#endif
+
+#include <fcntl.h>
+#include <WinSock2.h>
+#include <WS2tcpip.h>
+#include <ws2def.h>
+#include <ws2ipdef.h>
+#include <MSWSock.h>
+
+#include <folly/Portability.h>
+#include <folly/FilePortability.h>
+
+ // We don't actually support either of these flags
+ // currently.
+#define MSG_DONTWAIT 1
+#define MSG_EOR 0
+struct msghdr {
+  void* msg_name;
+  socklen_t msg_namelen;
+  struct iovec* msg_iov;
+  size_t msg_iovlen;
+  void* msg_control;
+  size_t msg_controllen;
+  int msg_flags;
+};
+
+#define SHUT_RD SD_RECEIVE
+#define SHUT_WR SD_SEND
+#define SHUT_RDWR SD_BOTH
+
+namespace folly { namespace socket_portability {
+  SOCKET fd_to_socket(int fd);
+  int socket_to_fd(SOCKET s);
+
+  // Unless you have a case where you would normally have
+  // to reference the function as being explicitly in the
+  // global scope, then you shouldn't be calling these directly.
+  int accept(int s, struct sockaddr* addr, int* addrlen);
+  int bind(int s, const struct sockaddr* name, int namelen);
+  int connect(int s, const struct sockaddr* name, int namelen);
+  int getpeername(int s, struct sockaddr* name, int* namelen);
+  int getsockname(int s, struct sockaddr* name, int* namelen);
+  int getsockopt(int s, int level, int optname, char* optval, int* optlen);
+  int getsockopt(int s, int level, int optname, void* optval, int* optlen);
+  int inet_aton(const char *cp, struct in_addr *inp);
+  const char* inet_ntop(int af, const void* src, char* dst, socklen_t size);
+  int listen(int s, int backlog);
+  int poll(pollfd* fds, int nfds, int timeout);
+  int recv(int s, char* buf, int len, int flags);
+  int recv(int s, void* buf, int len, int flags);
+  int recvfrom(int s, char* buf, int len, int flags, struct sockaddr* from, int* fromlen);
+  int recvfrom(int s, void* buf, int len, int flags, struct sockaddr* from, int* fromlen);
+  ssize_t recvmsg(int s, struct msghdr* message, int fl);
+  int send(int s, const char* buf, int len, int flags);
+  int send(int s, const void* buf, int len, int flags);
+  ssize_t sendmsg(int s, const struct msghdr *message, int fl);
+  int sendto(int s, const char* buf, int len, int flags, const sockaddr* to, int tolen);
+  int sendto(int s, const void* buf, int len, int flags, const sockaddr* to, int tolen);
+  // We use this first overload to convince MSVC to use our overloads, rather than the
+  // winsock ones, when optval is a const char*.
+  int setsockopt(int s, int level, int optname, const char* optval, int optlen);
+  int setsockopt(int s, int level, int optname, const void* optval, int optlen);
+  int shutdown(int s, int how);
+  int socketpair(int domain, int type, int protocol, int sockot[2]);
+
+  // This is the only function that _must_ be referenced via fsp::
+  // because there is no difference in parameter types to overload
+  // on.
+  int socket(int af, int type, int protocol);
+}}
+
+using namespace folly::socket_portability;
```

Comment:
Alright, will add the ignores, because that using is needed for this to play nice. Without that using, every single call to any of the socket portability functions would have to be prefixed with `fsp::`. In fact, that is how I originally wrote it, before getting tired of having to change everything and realizing I can just get the compiler to resolve to the correct overload of the function.